src/qibo/models/qcnn.py

```diff
@@ -0,0 +1,299 @@
+        c += self.one_qubit_unitary(bits[1], symbols[12:])
+
+        return c
+
+    def two_qubit_pool(self, source_qubit, sink_qubit, symbols):
+        """Make a circuit to do a parameterized 'pooling' operation, which
+        attempts to reduce entanglement down from two qubits to just one.
+        Symbols should be a list of 6 params.
+        """
+        pool_circuit = Circuit(self.nqubits)
+        sink_basis_selector = self.one_qubit_unitary(sink_qubit, symbols[0:3])
+        source_basis_selector = self.one_qubit_unitary(source_qubit, symbols[3:6])
+        pool_circuit += sink_basis_selector
+        pool_circuit += source_basis_selector
+        pool_circuit.add(gates.CNOT(source_qubit, sink_qubit))
+        pool_circuit += sink_basis_selector.invert()
+
+        return pool_circuit
+
+    def set_circuit_params(self, angles):
+
+        params = list(angles)
+        expanded_params = []
+        nbits = self.nqubits
+        for layer in range(self.nlayers):
+            nleft = nbits / (2**layer)
+            param_start = layer * self.nparams_layer
+            conv_params = params[param_start : param_start + self.nparams_conv]
+            pool_params = params[
+                param_start + self.nparams_conv : param_start + self.nparams_layer
+            ]
+            # pool_params += pool_params[2::-1]
+            pool_params += [-pool_params[2], -pool_params[1], -pool_params[0]]
+            expanded_params += conv_params * int(nleft if nleft > 2 else 1)
+            expanded_params += pool_params * int(nleft / 2)
+
+        self._circuit.set_parameters(expanded_params)
+
+    def Classifier_circuit(self, theta):
+        """
+        Args:
+            theta: list or numpy.array with the biases and the angles to be used in the circuit
+            nlayers: int number of layers of the varitional circuit ansatz
+            RY: if True, parameterized Rx,Rz,Rx gates are used in the circuit
+                if False, parameterized Ry gates are used in the circuit (default=False)
+        Returns:
+            Circuit implementing the variational ansatz for angles "theta"
+        """
+        bias = np.array(theta[0 : self.measured_qubits])
+        angles = theta[self.measured_qubits :]
+
+        self.set_circuit_params(angles)
+        return self._circuit
+
+    def Predictions(self, circuit, theta, init_state, nshots=10000):
+        """
+        Args:
+            theta: list or numpy.array with the biases to be used in the circuit
+            init_state: numpy.array with the quantum state to be classified
+            nshots: int number of runs of the circuit during the sampling process (default=10000)
+        Returns:
+            numpy.array() with predictions for each qubit, for the initial state
+        """
+        bias = np.array(theta[0 : self.measured_qubits])
+        circuit = circuit(init_state, nshots)
+        result = circuit.frequencies(binary=False)
```

Comment:
Another possibility here can be to use directly [`probabilities`](https://qibo.readthedocs.io/en/stable/api-reference/qibo.html#qibo.states.CircuitResult.probabilities), but both ways are effective.